js/src/hooks/useAutoCreateAdsMCAccounts.js

```diff
@@ -0,0 +1,202 @@
+
+	const {
+		hasFinishedResolution: hasFinishedResolutionForGoogleAdsAccount,
+		hasGoogleAdsConnection,
+	} = useGoogleAdsAccount();
+
+	const {
+		googleMCAccount,
+		hasFinishedResolution: hasFinishedResolutionForGoogleMCAccount,
+	} = useGoogleMCAccount();
+
+	const [ handleCreateAccount, { response } ] = useCreateMCAccount();
+	const [ upsertAdsAccount, { loading } ] = useUpsertAdsAccount();
+	const isGoogleMCConnected =
+		googleMCAccount?.status === GOOGLE_MC_ACCOUNT_STATUS.CONNECTED ||
+		googleMCAccount?.status === GOOGLE_MC_ACCOUNT_STATUS.INCOMPLETE;
+
+	const hasExistingMCAccount =
+		isGoogleMCConnected || existingMCAccounts?.length > 0;
+	const hasExistingAdsAccount =
+		hasGoogleAdsConnection || existingAdsAccount?.length > 0;
+
+	if (
+		initHasExistingMCAccountsRef.current === null &&
+		hasFinishedResolutionForExistingMCAccounts &&
+		hasFinishedResolutionForGoogleMCAccount
+	) {
+		initHasExistingMCAccountsRef.current = hasExistingMCAccount;
+	}
+
+	if (
+		initHasExistingAdsAccountsRef.current === null &&
+		hasFinishedResolutionForExistingAdsAccount &&
+		hasFinishedResolutionForGoogleAdsAccount
+	) {
+		initHasExistingAdsAccountsRef.current = hasExistingAdsAccount;
+	}
+
+	const createMCAccount = useCallback( async () => {
+		await handleCreateAccount();
+	}, [ handleCreateAccount ] );
+
+	const createAdsAccount = useCallback( async () => {
+		await upsertAdsAccount();
+	}, [ upsertAdsAccount ] );
+
+	const createBothAccounts = useCallback( async () => {
+		await createMCAccount();
+		await createAdsAccount();
+	}, [ createMCAccount, createAdsAccount ] );
+
+	const accountCreationChecksResolved =
+		initHasExistingAdsAccountsRef.current !== null &&
+		initHasExistingMCAccountsRef.current !== null;
+
+	const shouldCreateAdsAccount =
+		initHasExistingAdsAccountsRef.current === false &&
+		initHasExistingMCAccountsRef.current === true;
+
+	const shouldCreateMCAccount =
+		initHasExistingAdsAccountsRef.current === true &&
+		initHasExistingMCAccountsRef.current === false;
+
+	const shouldCreateBothAccounts =
+		! initHasExistingAdsAccountsRef.current &&
+		! initHasExistingMCAccountsRef.current;
+
+	const isCreatingAccounts =
+		isCreatingAdsAccountsRef.current ||
+		isCreatingMCAccountsRef.current ||
+		isCreatingBothAccountsRef.current;
+
+	useEffect( () => {
+		// Ads account check
+		if ( isCreatingAdsAccountsRef.current === true && ! loading ) {
+			isCreatingAdsAccountsRef.current = false;
+			accountsCreatedRef.current = true;
+		}
+
+		// MC account check
+		if (
+			isCreatingMCAccountsRef.current === true &&
+			response?.status === 200
+		) {
+			isCreatingMCAccountsRef.current = false;
+			accountsCreatedRef.current = true;
+		}
+
+		// both accounts check
+		if (
+			isCreatingBothAccountsRef.current === true &&
+			response?.status === 200 &&
+			! loading
+		) {
+			isCreatingBothAccountsRef.current = false;
+			accountsCreatedRef.current = true;
+		}
+	}, [ response, loading ] );
```

Comment:
IMO, it is safe to keep them as is because whether the accounts have been created or not depends on `response` and `loading` values, thus we are using `useEffect` here.

Maybe, we can further refactor this part in #2582 where we may want to use `useApiFetchCallback` directly to create ads account in order to avoid unnecessary notices during ads creation.

Reply:
Removed this ref. Now, `useAutoCreateAdsMCAccounts` hook uses only one ref `shouldCreateAccounts` internally to check whether we need to trigger any of the account creation process.